lib/actions/auth.ts

```diff
@@ -0,0 +1,209 @@
+limitations under the License.
+*/
+
+import { CommandDefinition } from 'capitano';
+
+export const login: CommandDefinition<
+	{},
+	{
+		token: string | boolean;
+		web: boolean;
+		credentials: boolean;
+		email: string;
+		password: string;
+	}
+> = {
+	signature: 'login',
+	description: 'login to balena',
+	help: `\
+Use this command to login to your balena account.
+
+This command will prompt you to login using the following login types:
+
+- Web authorization: open your web browser and prompt you to authorize the CLI
+from the dashboard.
+
+- Credentials: using email/password and 2FA.
+
+- Token: using a session token or API key from the preferences page.
+
+Examples:
+
+	$ balena login
+	$ balena login --web
+	$ balena login --token "..."
+	$ balena login --credentials
+	$ balena login --credentials --email johndoe@gmail.com --password secret\
+`,
+	options: [
+		{
+			signature: 'token',
+			description: 'session token or API key',
+			parameter: 'token',
+			alias: 't',
+		},
+		{
+			signature: 'web',
+			description: 'web-based login',
+			boolean: true,
+			alias: 'w',
+		},
+		{
+			signature: 'credentials',
+			description: 'credential-based login',
+			boolean: true,
+			alias: 'c',
+		},
+		{
+			signature: 'email',
+			parameter: 'email',
+			description: 'email',
+			alias: ['e', 'u'],
+		},
+		{
+			signature: 'password',
+			parameter: 'password',
+			description: 'password',
+			alias: 'p',
+		},
+	],
+	primary: true,
+	async action(_params, options, done) {
+		type Options = typeof options;
+		const _ = await import('lodash');
+		const Promise = await import('bluebird');
+		const balena = (await import('balena-sdk')).fromSharedOptions();
+		const auth = await import('../auth');
+		const form = await import('resin-cli-form');
+		const patterns = await import('../utils/patterns');
+		const messages = await import('../utils/messages');
+
+		const doLogin = (loginOptions: Options): Promise<void> => {
+			if (loginOptions.token != null) {
+				return Promise.try(() => {
+					if (_.isString(loginOptions.token)) {
+						return loginOptions.token;
+					}
+					return form.ask({
+						message: 'Session token or API key from the preferences page',
+						name: 'token',
+						type: 'input',
+					});
+				})
+					.then(balena.auth.loginWithToken)
+					.tap(() =>
+						balena.auth.whoami().then(username => {
+							if (!username) {
+								return patterns.exitWithExpectedError(
+									'Token authentication failed',
+								);
+							}
+						}),
+					);
+			} else if (loginOptions.credentials) {
+				return patterns.authenticate(loginOptions);
+			} else if (loginOptions.web) {
+				console.info('Connecting to the web dashboard');
+				return auth.login().return();
+			}
+
+			return patterns.askLoginType().then(loginType => {
+				if (loginType === 'register') {
+					const signupUrl = 'https://dashboard.balena-cloud.com/signup';
+					require('opn')(signupUrl, { wait: false });
```

Comment:
hmm maybe it's OK, but the original code here was:
```
require('open')(signupUrl, { url: true, wait: false })
```
note 'opn' _vs. 'open', and the `'url'` attribute.

Reply:
Yeah I haven't rebased this on my other changes yet - this will definitely need updating (and can also become `import()` with the added types from `open`)

Reply:
done